sdk/lib/opentelemetry/sdk/trace/tracer.rb

```diff
@@ -56,7 +54,8 @@ def internal_create_span(result, name, kind, trace_id, span_id, parent_span_id,
             attributes = attributes&.merge(result.attributes) || result.attributes
             active_trace_config = OpenTelemetry.tracer_provider.active_trace_config
             active_span_processor = OpenTelemetry.tracer_provider.active_span_processor
-            Span.new(context, name, kind, parent_span_id, active_trace_config, active_span_processor, attributes, links, start_timestamp || Time.now, @resource, @instrumentation_library)
+            resource = OpenTelemetry.tracer_provider.resource
+            Span.new(context, name, kind, parent_span_id, active_trace_config, active_span_processor, attributes, links, start_timestamp || Time.now, resource, @instrumentation_library)
```

Comment:
I *think* this is incorrect. Or at least unnecessary. We should remove the `resource` parameter -- for now, can you just pass `nil`?

Reply:
What the issue here exactly? A span should have a reference to a resource.

Reply:
A span needs a reference to an `InstrumentationLibrary`. The `Resource` is global (a singleton). I don’t see the need to attach the same Resource reference to every span when it is always available from `OpenTelemetry.tracer_provider.resource`.

Reply:
Unfortunately, that's a current design flaw of the Ruby library. Technically, multiple TracerProviders should be able to exist in a single process. They should be able to have different resources, but share the same export pipeline. So, the resource is not exactly a global singleton. This was discussed somewhat in https://github.com/open-telemetry/opentelemetry-specification/issues/508 and I am supposed to be writing some clarifying language for the spec about this topic.

For the Ruby SDK. I think it's fine to have a single TracerProvider configured by default, as long as we allow additional ones to be instantiated. We'll need to make sure that references go through TracerProvider instances, rather than the classes. There are few places we were need to clean up static references in this regard.

Reply:
I'll try to get some more clarification about the multiple TracerProvider situation at the spec SIG tomorrow. I've heard that it's a use case, but given that you can only have one global TracerProvider at the API level, I'm not really sure how this works in practice, or if it's necessarily a use case that needs to be preserved.

Reply:
This is the relevant portion of the spec: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracerprovider

> Normally, the TracerProvider is expected to be accessed from a central place. Thus, the API SHOULD provide a way to set/register and access a global default TracerProvider.

> Notwithstanding any global TracerProvider, some applications may want to or have to use multiple TracerProvider instances, e.g. to have different configuration (like SpanProcessors) for each (and consequently for the Tracers obtained from them), or because its easier with dependency injection frameworks. Thus, implementations of TracerProvider SHOULD allow creating an arbitrary number of TracerProvider instances.

Noting that the spec lists this as a SHOULD, not a MUST.

Reply:
If I'm following correctly, for the `sdk/lib/opentelemetry/sdk/trace/tracer.rb` code could we not "just" pass in a reference to the `tracer_provider` that created it?  The `tracer_provider` is accessed in multiple places within the `tracer.rb` class already, why not pass it in as part of the initializer for the tracer?

That would solve the issue of supporting multiple `tracer_provider`s for this particular code path, no?

It would also remove some of the implicit coupling that the tracer has on the tracer_provider.
https://github.com/robertlaurin/opentelemetry-ruby/blob/cf02a49c38e230f805dcfd9bbb39b8b9a8caabfb/sdk/lib/opentelemetry/sdk/trace/tracer.rb#L27

Reply:
You're following correctly @robertlaurin. I think the core issue is that because of the multiple TracerProvider use case, each span needs a reference to a resource. Ultimately, this is because TracerProviders can be initialized with different resources, but share the same export pipeline. As you point out, we can support this use case pretty easily passing a TracerProvider to a Tracer during initialization and maintaining a reference to it.

Reply:
I'm not sure how much of this needs to be addressed in this PR. It would be fine to create issues for additional work that isn't directly related.

Reply:
I agree that I don't want to grow this PR into additional work that isn't directly related. 
 I think this change here https://github.com/open-telemetry/opentelemetry-ruby/pull/263/commits/a43ce3a580aa10b3f34e5cda80915e42412ceffc leaves the code in a good places to be improved on.